src/map/guild.cpp

```diff
@@ -880,62 +927,95 @@ int guild_member_added(int guild_id,uint32 account_id,uint32 char_id,int flag) {
 /*==========================================
  * Player request leaving a given guild_id
  *----------------------------------------*/
-int guild_leave(struct map_session_data* sd, int guild_id, uint32 account_id, uint32 char_id, const char* mes) {
-	struct guild *g;
+bool guild_leave(struct map_session_data* sd, int guild_id, uint32 account_id, uint32 char_id, const char* mes) {
+	nullpo_retr( false, sd );
 
-	nullpo_ret(sd);
+	struct guild *g = sd->guild;
 
-	g = sd->guild;
+	if( g == nullptr ){
+		return false;
+	}
 
-	if(g==NULL)
-		return 0;
+	if( g->instance_id > 0 && battle_config.instance_block_leave ){
+		clif_displaymessage( sd->fd, msg_txt( sd, 228 ) ); //TODO: message
+		return false;
+	}
+
+	if( map_getmapflag( sd->bl.m, MF_GUILDLOCK) ){
+		clif_displaymessage( sd->fd, msg_txt( sd, 228 ) ); //Guild modification is disabled on this map.
+		return false;
+	}
+
+	if( sd->bg_id ){
+		clif_displaymessage( sd->fd, msg_txt( sd, 670 ) ); //"You can't leave battleground guilds."
+		return false;
+	}
 
 	if(sd->status.account_id!=account_id ||
 		sd->status.char_id!=char_id || sd->status.guild_id!=guild_id ||
 		map_flag_gvg2(sd->bl.m))
-		return 0;
+		return false;
 
 	guild_trade_bound_cancel(sd);
-	intif_guild_leave(sd->status.guild_id, sd->status.account_id, sd->status.char_id,0,mes);
-	return 0;
+
+	return intif_guild_leave( sd->status.guild_id, sd->status.account_id, sd->status.char_id, 0, mes );
 }
 
 /*==========================================
  * Request remove a player to a given guild_id
  *----------------------------------------*/
-int guild_expulsion(struct map_session_data* sd, int guild_id, uint32 account_id, uint32 char_id, const char* mes) {
+bool guild_expulsion(struct map_session_data* sd, int guild_id, uint32 account_id, uint32 char_id, const char* mes) {
 	struct map_session_data *tsd;
-	struct guild *g;
-	int i,ps;
 
-	nullpo_ret(sd);
+	nullpo_retr( false, sd );
 
-	g = sd->guild;
+	struct guild *g = sd->guild;
 
-	if(g==NULL)
-		return 0;
+	if( g == nullptr ){
+		return false;
+	}
 
-	if(sd->status.guild_id!=guild_id)
-		return 0;
+	if( sd->status.guild_id != guild_id ){
+		return false;
+	}
 
-	if( (ps=guild_getposition(sd))<0 || !(g->position[ps].mode&GUILD_PERM_EXPEL) )
-		return 0;	//Expulsion permission
+	if( !guild_has_permission( sd, GUILD_PERM_EXPEL ) ){
+		return false;
+	}
+
+	if( g->instance_id > 0 && battle_config.instance_block_expulsion ){
+		return false;
+	}
+
+	// TODO: for leave this is different messages
+	if( map_getmapflag( sd->bl.m, MF_GUILDLOCK ) || sd->bg_id ){
```

Comment:
`if(sd->bg_id || map_getmapflag( sd->bl.m, MF_GUILDLOCK )`
would be faster, there is no full elevation in c++